iso-image: Fix issues with UEFI installer image

samueldr opened the pull request:
Tries to handle #123376

###### Things done

Verified on:

 - qemu (-cdrom)
 - Acer c720p
 - Pinebook (A64)

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

samueldr committed:
iso-image: unqualified root → ($root)

This technically changes nothing. In practice `$root` is always the
"CWD", whether searched for automatically or not.

But this serves to announce we are relying on `$root`... I guess...

samueldr committed:
iso-image: Force gfxmode

https://www.gnu.org/software/grub/manual/grub/html_node/gfxmode.html

samueldr committed:
iso-image: change date on all files

It may be that in some conditions dates earlier than 1980 on FAT on GRUB
2.06~ish will cause failures

https://github.com/NixOS/nixpkgs/issues/123376#issuecomment-845515035

ncfavier commented on nixos/modules/installer/cd-dvd/iso-image.nix:
```diff
@@ -403,7 +421,9 @@ let
-      touch --date=@0 ./EFI ./boot
+
+      # Rewrite dates for everything in the FS
+      find . -exec touch --date=@$DATE '{}' ';'
```
You could just use `--date=2000-01-01` here to avoid magic numbers.

samueldr replied:
TBF a magic number is magic when used with neither name or a description :).

But yes, much preferable this way.

ncfavier commented on nixos/modules/installer/cd-dvd/iso-image.nix:
```diff
@@ -403,7 +421,9 @@ let
-      touch --date=@0 ./EFI ./boot
+
+      # Rewrite dates for everything in the FS
+      find . -exec touch --date=@$DATE '{}' ';'
```
```suggestion
      find . -exec touch --date=2000-01-01 {} +
```

to process more than one file at once

hyperfekt commented:
The date issue is fixed, the one of failed theming is not. I can't commit on the resolution one since I don't have that.

hyperfekt commented:
@ncfavier Can you confirm the font problem is fixed? In that case we should merge a PR with either just the `(hd0) → ($root)` commit or that *and* the revert backed out and resolve the blocker.

ncfavier commented:
Building the ISO now.

To be clear, `($root)/foo` is the same as just `/foo`. I don't think we should merge this until we figure out a way to properly set the `$root`, which probably involves using the `search` command.

hyperfekt commented:
Well, that's just an issue of lacking theming, and the other commits prevent booting. Which is why, as I see it, the latter should prevent a release of 21.05 of while the former should not. This is why I'd like to separate the two parts and merge them separately, so we can take our time and find out how to solve the theming issue on all platforms without making @jonringer sad.

ncfavier commented:
Fair enough.

jonringer commented:
> without making @jonringer sad.

at this point, all software makes me sad

ncfavier commented:
OK, I can confirm that this fixes the date bug and the resolution bug (at least on my machine).

d64034d5e79cf454a995d3270931d02c67287e0b and d64034d5e79cf454a995d3270931d02c67287e0b cancel each other out, so you might want to rebase them out.

For the record, [here's a list](https://en.wikipedia.org/wiki/VESA_BIOS_Extensions#Other_commonly_available_graphics_modes) of common VBE video modes.

hyperfekt commented:
Fix is still good after the force-push.

samueldr committed:
iso-image: Improve disk detection

This should help in rare hardware-specific situations where the root is
not automatically detected properly.

We search using a marker file. This should help some weird UEFI setups
where the root is set to `(hd0,msdos2)` by default.

Defaulting to `(hd0)` by looking for the ESP **will break themeing**. It
is unclear why, but files in `(hd0,msdos2)` are not all present as they
should be.

This also fixes an issue introduced with cb5c4fcd3c5d4070f040d591b2dd1da580f234d1
where rEFInd stopped booting in many cases. This is because it ended up
using (hd0) rather than using the `search` which was happening
beforehand, which in turn uses (hd0,msdos2), which is the ESP.
Putting back the `search` here fixes that.

samueldr commented:
AFAIUI `vbe` is irrelevant here since we're in an UEFI gop driver.

ncfavier commented:
From the manual:

> Note that you can only use modes which your graphics card supports via VESA BIOS Extensions (VBE)

I have no idea though.

samueldr commented:
Yeah, GRUB's documentation is still heavily biased to their "legacy" program.

We're exclusively using the UEFI version of GRUB with the iso.

ncfavier commented:
Aha you're right, `videoinfo` lists modes that aren't in that table.

samueldr commented:
haha! detecting as `(hd0,msdos2)` breaks themes here! An artificial reproducer!

So I guess the issue was that GRUB detected itself as `(hd0,msdos2)`. (For the theme issue.)

hyperfekt commented:
The latest commit is unsuccessful in fixing the theming issue for me, in case that is what it is supposed to do.

samueldr commented:
@hyperfekt yeah, I pushed it eagerly, and then I figured out that reproduces the themeing issue.

samueldr commented:
I need to test on `AArch64` still.

hyperfekt commented:
This is the one. Theming works now, and the date thing is still resolved. Good job!

jonringer commented:
awesome turn around. Thanks for working on this all :)

samueldr commented:
Verified with *[Tow-Boot](https://tow-boot.org/)* on my Pinebook A64 that it works just as well as it did.

jonringer commented:
@GrahamcOfBorg test grub

jonringer reviewed (approved):
diff LGTM

tests still checkout on x86, aarch64 was tested

ncfavier commented:
This works for me, but only by accident: the `nixos-installer-image` marker file ends up on both `(hd0)` and `(hd0,msdos2)`, so whichever GRUB tries first is picked. We should make sure to use a file that's only on `(hd0)`.

samueldr commented:
It's not actually a different FS. It's `isohybrid` trickery.

We should look into why some files are not present.

But for the time being the failure state is seemingly lack of theme. Which isn't much of an issue as long as the user can boot in the system!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/21-05-has-been-released/13407/5

codelamb commented:
Hi,  

I have noticed that  iso-image plasma5-20.09.2538.0cfd08f4881-x86_64-linux.iso  probably lacks tools for dealing with fat32 filesystems. I had issues installing Nixos 21.05 on Samsung NP530U4E. System was damaged while upgrading from 20.03 to 21.05, giving the error "unable to build, disc was probably full" (quoting from memory, sorry). After trying to reinstall from USB got the same error message. Gparted was showing error message, that system lacks tools to deal with fat32 filesystems.I assume that  nixos-install was probably unable to detect fat32 boot partition and was assuming that disk is full. I was strictly following [official manual for UEFI systems](https://nixos.org/manual/nixos/stable/index.html#sec-installation-partitioning-UEFI). I was able to install Nixos only after manual installation of mtools with `nix-env -f '<nixpkgs>' -iA mtools`. Just in case I have also declared mtools in configuration.nix

samueldr commented:
(I "moved" this comment into a new issue as it is irrelevant to this PR see #132094 )
